internal/app/machined/pkg/controllers/files/nqn.go

```diff
@@ -0,0 +1,119 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package files
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/google/uuid"
+	"github.com/siderolabs/gen/optional"
+	"go.uber.org/zap"
+
+	runtimetalos "github.com/siderolabs/talos/internal/app/machined/pkg/runtime"
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+	"github.com/siderolabs/talos/pkg/machinery/resources/files"
+)
+
+// NQNController creates an EtcFileSpec for the NVMe Qualified Name (NQN) and HostID file.
+type NQNController struct {
+	V1Alpha1Mode runtimetalos.Mode
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *NQNController) Name() string {
+	return "files.NQNController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *NQNController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: files.NamespaceName,
+			Type:      files.EtcFileStatusType,
+			ID:        optional.Some("machine-id"),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *NQNController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: files.EtcFileSpecType,
+			Kind: controller.OutputShared,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *NQNController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	// Skip the controller if we're running in a container.
+	if ctrl.V1Alpha1Mode == runtimetalos.ModeContainer {
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		// wait for machine-id to be available
+		_, err := safe.ReaderGetByID[*files.EtcFileStatus](ctx, r, "machine-id")
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("failed to get machine-id etcfile status: %w", err)
+		}
+
+		machineID, err := os.ReadFile("/etc/machine-id")
+		if err != nil {
+			return fmt.Errorf("failed to read machine-id: %w", err)
+		}
+
+		hostID, err := uuid.FromBytes(machineID[:16])
+		if err != nil {
+			return fmt.Errorf("failed to convert machine-id to UUID: %w", err)
+		}
+
+		if err := safe.WriterModify(ctx, r, files.NewEtcFileSpec(files.NamespaceName, "nvme/hostid"),
+			func(r *files.EtcFileSpec) error {
+				spec := r.TypedSpec()
+
+				spec.Contents = []byte(hostID.String())
```

Comment:
uuid is actually []byte underneath, so you can do a simple assign here I think. Or []byte(hostID)

Reply:
Oh, yeah, my bad. We need string representation stored in byte slice.